src/core/lib/security/authorization/sdk_server_authz_filter.cc

```diff
@@ -0,0 +1,224 @@
+    GRPC_CLOSURE_INIT(&recv_initial_metadata_ready_, RecvInitialMetadataReady,
+                      elem, grpc_schedule_on_exec_ctx);
+    GRPC_CLOSURE_INIT(&recv_trailing_metadata_ready_, RecvTrailingMetadataReady,
+                      elem, grpc_schedule_on_exec_ctx);
+  }
+
+  ~CallData() { GRPC_ERROR_UNREF(recv_initial_metadata_error_); }
+
+  static void StartTransportStreamOpBatch(
+      grpc_call_element* elem, grpc_transport_stream_op_batch* batch);
+  static grpc_error_handle Init(grpc_call_element* elem,
+                                const grpc_call_element_args* args);
+  static void Destroy(grpc_call_element* elem,
+                      const grpc_call_final_info* /*final_info*/,
+                      grpc_closure* /*ignored*/);
+
+ private:
+  static void RecvInitialMetadataReady(void* arg, grpc_error_handle error);
+  static void RecvTrailingMetadataReady(void* user_data, grpc_error_handle err);
+
+  grpc_core::CallCombiner* call_combiner_;
+  grpc_transport_stream_op_batch* recv_initial_metadata_batch_;
+  grpc_closure* original_recv_initial_metadata_ready_;
+  grpc_closure recv_initial_metadata_ready_;
+  grpc_error_handle recv_initial_metadata_error_ = GRPC_ERROR_NONE;
+  grpc_closure* original_recv_trailing_metadata_ready_;
+  grpc_closure recv_trailing_metadata_ready_;
+  grpc_error_handle recv_trailing_metadata_error_;
+  bool seen_recv_trailing_metadata_ready_ = false;
+};
+
+bool IsAuthorized(ChannelData* chand, grpc_transport_stream_op_batch* batch) {
+  grpc_core::EvaluateArgs args(
+      batch->payload->recv_initial_metadata.recv_initial_metadata,
+      &chand->channel_args_);
+  if (chand->provider_->deny_engine() != nullptr) {
+    grpc_core::AuthorizationEngine::Decision decision =
+        chand->provider_->deny_engine()->Evaluate(args);
+    if (decision.type ==
+        grpc_core::AuthorizationEngine::Decision::Type::kDeny) {
+      gpr_log(GPR_DEBUG, "Request denied. Matching policy name: %s.",
+              decision.matching_policy_name.c_str());
+      return false;
+    }
+  }
+  if (chand->provider_->allow_engine() != nullptr) {
+    grpc_core::AuthorizationEngine::Decision decision =
+        chand->provider_->allow_engine()->Evaluate(args);
+    if (decision.type ==
+        grpc_core::AuthorizationEngine::Decision::Type::kAllow) {
+      gpr_log(GPR_DEBUG, "Request allowed. Matching policy name: %s.",
+              decision.matching_policy_name.c_str());
+      return true;
+    }
+  }
+  gpr_log(GPR_DEBUG, "Request denied. No match found.");
+  return false;
+}
+
+void CallData::RecvInitialMetadataReady(void* arg, grpc_error_handle error) {
+  grpc_call_element* elem = static_cast<grpc_call_element*>(arg);
+  ChannelData* chand = static_cast<ChannelData*>(elem->channel_data);
+  CallData* calld = static_cast<CallData*>(elem->call_data);
+  grpc_transport_stream_op_batch* batch = calld->recv_initial_metadata_batch_;
+  if (error == GRPC_ERROR_NONE) {
+    if (!IsAuthorized(chand, batch)) {
```

Comment:
@markdroth 
I had a doubt here. So if the request is authorized, we go ahead and execute the next filter right?
For ex. in xds enabled servers, where we may have xds filters after sdk filter in filter chain.

Reply:
Are we supporting the use-case where we also have an xDS supplied RBAC policy in addition to the SDK RBAC policy on the same server?